src/main/kotlin/cacophony/grammar/syntaxtree/Expression.kt

```diff
@@ -0,0 +1,279 @@
+package cacophony.grammar.syntaxtree
+
+import cacophony.utils.Location
+
+sealed class Expression(val range: Pair<Location, Location>) { // everything in cacophony is an expression
+    // artificial instance, can be useful when calculating values of nested expressions
+    class EmptyExpression(
+        range: Pair<Location, Location>,
+    ) : Expression(range)
+
+    class Variable(
+        range: Pair<Location, Location>,
+    ) : Expression(range)
+
+    // should Type also inherit expression? if so, can we do for example "; Bool;" in our language?
+    class Type(
+        range: Pair<Location, Location>,
+    ) : Expression(range)
+
+    class VariableDeclaration(
+        range: Pair<Location, Location>,
+        val keywordLet: Keyword.KeywordLet,
+    ) : Expression(range)
+
+    class FunctionArgument(
+        range: Pair<Location, Location>,
+        type: Type,
+    ) : Expression(range)
+
+    class VariableTypedDeclaration(
+        range: Pair<Location, Location>,
+        variable: VariableDeclaration,
+        type: Type,
+    ) : Expression(range)
+
+    class Literal(
+        range: Pair<Location, Location>,
+    ) : Expression(range)
+
+    class Keyword(
+        range: Pair<Location, Location>,
+    ) : Expression(range) {
+        class KeywordBreak(
+            range: Pair<Location, Location>,
+        ) : Expression(range)
+
+        // it is possible that following keywords are unnecessary, in this case break would be probably moved to statement
+        class KeywordReturn(
+            range: Pair<Location, Location>,
+        ) : Expression(range)
+
+        class KeywordLet(
+            range: Pair<Location, Location>,
+        ) : Expression(range)
+
+        class KeywordIf(
+            range: Pair<Location, Location>,
+        ) : Expression(range)
+
+        class KeywordThen(
+            range: Pair<Location, Location>,
+        ) : Expression(range)
+
+        class KeywordElse(
+            range: Pair<Location, Location>,
+        ) : Expression(range)
+
+        class KeywordWhile(
+            range: Pair<Location, Location>,
+        ) : Expression(range)
+
+        class KeywordDo(
+            range: Pair<Location, Location>,
+        ) : Expression(range)
+    }
+
+    // expression in parenthesis
+    class NestedExpression(
```

Comment:
**suggestion:** maybe renaming this to `ParenthesisGroup` or would be more explicit (also `Expression.NestedExpression` repeats `Expression` )